Sources/Sentry/SentryFileManager.m

```diff
@@ -217,44 +225,72 @@ - (void)handleEnvelopesLimit
 {
     NSData *sessionData = [SentrySerialization dataWithSession:session error:nil];
-    [SentryLog logWithMessage:[NSString stringWithFormat:@"Writing to current session: %@",
-                                        self.currentSessionFilePath]
+    [SentryLog logWithMessage:[NSString stringWithFormat:@"Writing session: %@", sessionFilePath]
                      andLevel:kSentryLogLevelDebug];
-    @synchronized(self.currentSessionFilePath) {
-        [sessionData writeToFile:self.currentSessionFilePath options:NSDataWritingAtomic error:nil];
+    @synchronized(_sessionLock) {
+        [sessionData writeToFile:sessionFilePath options:NSDataWritingAtomic error:nil];
     }
 }
 
 - (void)deleteCurrentSession
 {
-    [SentryLog logWithMessage:[NSString stringWithFormat:@"Deleting current session: %@",
-                                        self.currentSessionFilePath]
+    [self deleteSession:self.currentSessionFilePath];
+}
+
+- (void)deleteCrashedSession
+{
+    [self deleteSession:self.crashedSessionFilePath];
+}
+
+- (void)deleteSession:(NSString *)sessionFilePath
+{
+    [SentryLog logWithMessage:[NSString stringWithFormat:@"Deleting session: %@", sessionFilePath]
                      andLevel:kSentryLogLevelDebug];
     NSFileManager *fileManager = [NSFileManager defaultManager];
-    @synchronized(self.currentSessionFilePath) {
-        [fileManager removeItemAtPath:self.currentSessionFilePath error:nil];
+    @synchronized(_sessionLock) {
+        [fileManager removeItemAtPath:sessionFilePath error:nil];
     }
 }
 
 - (SentrySession *_Nullable)readCurrentSession
 {
-    [SentryLog logWithMessage:[NSString stringWithFormat:@"Reading from current session: %@",
-                                        self.currentSessionFilePath]
-                     andLevel:kSentryLogLevelDebug];
+    return [self readSession:self.currentSessionFilePath];
+}
+
+- (SentrySession *_Nullable)readCrashedSession
+{
+    return [self readSession:self.crashedSessionFilePath];
+}
+
+- (SentrySession *)readSession:(NSString *)sessionFilePath
+{
+    [SentryLog
+        logWithMessage:[NSString stringWithFormat:@"Reading from session: %@", sessionFilePath]
+              andLevel:kSentryLogLevelDebug];
     NSFileManager *fileManager = [NSFileManager defaultManager];
-    NSData *currentSessionData = nil;
-    @synchronized(self.currentSessionFilePath) {
-        currentSessionData = [fileManager contentsAtPath:self.currentSessionFilePath];
-        if (nil == currentSessionData) {
+    NSData *currentData = nil;
+    @synchronized(_sessionLock) {
+        currentData = [fileManager contentsAtPath:sessionFilePath];
+        if (nil == currentData) {
             return nil;
         }
-        SentrySession *currentSession = [SentrySerialization sessionWithData:currentSessionData];
+        SentrySession *currentSession = [SentrySerialization sessionWithData:currentData];
```

Comment:
L: You could end the sync block before this, serialization can happen outside of the lock.

Reply:
Yep makes sense. Was like this, good catch 💯